### ISSUE TITLE
Replacing timeout in Net::HTTP#connect with nonblocking io

### ISSUE THREAD

**Issue by mohamedhafez** [jruby/jruby#6523]
Years ago, I submitted a patch to MRI ruby to replace the timeout in Net::HTTP#connect with nonblocking io, however it was [shot down](https://bugs.ruby-lang.org/issues/12435) because there was no good way to time out the DNS call in `Socket.sockaddr_in` due to limitations in the C libraries available.

Are we subject to the same limitations in Java? I've been running a version of [this patch](https://github.com/jruby/jruby/compare/master...mohamedhafez:patch-1?diff=split) for years because I was terrified by @headius's [blog post from way back in 2008](http://blog.headius.com/2008/02/ruby-threadraise-threadkill-timeoutrb.html) about the dangers of ruby's `timeout`, lol.

So my questions are:

a) Is it possible to introduce a timeout for `Socket.sockaddr_in` in the Java implementation?

b) How big of a deal is this / is this worth the effort to address? With the use of `AtomicBoolean` and other synchronization methods and a thread pool executor in Jruby's `Timeout.timeout` implementation, are the concerns of that old blog post relevant to Net::HTTP#connect's use of `Timeout.timeout` in JRuby today?

**Comment by mohamedhafez** [comment]
Actually, it looks like a way simpler patch would fix this: https://github.com/jruby/jruby/compare/master...mohamedhafez:patch-2?diff=split

While this still has the issue of not accounting for DNS timeouts, it looks like in Ruby 2.7 `Socket.tcp` does take a `:resolv_timeout` option, and then with just a 3 line patch we'd be able to do away with `Timeout.timeout`.

So I guess to revise my question, once JRuby is compatible with 2.7.0, what do you guys think about incorporating a patch like this, including a `:resolv_timeout`, into JRuby's version of net/http.rb? Or is the use of `Timeout.timeout` no longer an issue with the way JRuby implements it?

**Comment by headius** [comment]
The big question here will be if there is a way to tell Java to timeout the DNS resolution, since we currently implement all the TCP sockets using Java's socket libraries. Maybe you can do some research on that?

**Comment by mohamedhafez** [comment]
Hmm:/ So it looks like Java's InetAddress#getByName​, doesn't take a timeout option, and [just ends up calling getaddrinfo under the hood](https://medium.com/@maheshsenni/host-name-resolution-in-java-80301fea465a), at least in OpenJDK.

Trying to read up on alternatives, one suggestion was offloading the call to InetAddress#getByName to a thread pool executor, and then waiting on the `Future` with a timeout, but then we have one DNS lookup per thread. Maybe that's not as bad as it sounds, like if we can't reach the DNS server and the requests are taking forever, they'd all end up timing out any way so it doesn't matter if they are waiting in a queue I guess, no? Not ideal though, I know.

There's some 3rd party libraries that do implement async DNS lookups I believe. I think [Netty's DNSNameResolver](https://netty.io/4.1/api/io/netty/resolver/dns/DnsNameResolver.html) would do it, as well as [Akka's async-dns](https://doc.akka.io/docs/akka/current/io-dns.html). Not sure if those are too heavyweight or pull in too many dependencies for JRuby or not.

@headius how much of a concern is the use of `Timeout.timeout` in JRuby these days? I looked through the source code of the JRuby implementation, and with the use of the `ScheduledThreadPoolExecutor` it shouldn't be as inefficient as spinning up a new thread for each timeout, and thinking it through I don't *think* the JRuby implementation is susceptible to the race conditions you described in the old blog post... am I correct in that assessment or is there something I haven't thought of?

**Comment by mohamedhafez** [comment]
So it turns out that C-Ruby's [Timeout::timeout can't time out the getaddrinfo C call in the first place](https://bugs.ruby-lang.org/issues/17528)! So the status quo on both JRuby and C-Ruby is that the `open_timeout` option has no effect on the DNS lookup anyway, both have to wait around for the `getaddrinfo` C call they both end up calling to time out on its own (30s on MacOS 11.1 for example).

Considering that `Net::HTTP` was moved to a gem that any JRuby user should be able to pull in (once JRuby is compatible with 2.6 anyway), I figure this is probably better dealt with upstream and have [opened a new issue there and offered a fix](https://github.com/ruby/net-http/issues/6).

I'll go ahead and close this issue here, but just wanted to suggest first that perhaps the couple JRuby-specific changes to `Net::HTTP` should be be added to a Java version of the new [net-http](https://github.com/ruby/net-http) gem?

**Comment by headius** [comment]
@mohamedhafez We would LOVE to run off the gem but as you point out we have a few JRuby-specific tweaks in there. The changes should mostly be the extra handling of JDK proxy settings, I think? Maybe we can figure out a less invasive way to add this behavior?

I agree with fixing this upstream but for JRuby 9.3 we will also want to make that change locally (net/http is not a gem in Ruby 2.6; we might be able to get away with it, but had not planned to do so).

**Comment by headius** [comment]
> Are we subject to the same limitations in Java? I've been running a version of this patch for years because I was terrified by @headius's blog post from way back in 2008 about the dangers of ruby's timeout, lol.

I realized I never actually answered this, or addressed the use of timeout here in JRuby.

First off, I want to make it clear that JRuby does *not* spin up a thread for every timeout call **unless the timeout is exceeded**. We use a "scheduled" thread pool to avoid the overhead of the thread unless necessary. Since timeout is usually considered an error or failure, the vast majority of timeout calls will never spin up an additional thread on JRuby.

Second, the issues in my 2008 post are still there, since whenever the timeout **does** fire it will need to asynchronously interrupt the executing thread. The issues I described have been addressed by using the `Thread.handle_interrupt` feature to disable those interrupts in critical sections, but that must be done everywhere by hand and there's a lot of code out there still not doing it (or not doing it right). See #6527 for an example of the right `handle_interrupt` pattern, and a bug caused by the wrong pattern.

The bottom line is that while lower overhead on JRuby, timeout still suffers from the same issues I described in 2008 unless you explicitly mitigate those issues, and using API-supported timeout mechanisms (like nonblock etc) will always be better and safer.

**Comment by headius** [comment]
Looks like this is the PR that added interruptible `getaddrinfo` to CRuby using `getaddrinfo_a`: https://github.com/ruby/ruby/pull/2827

It does not appear to have been in a release until 3.0, is that correct?

I notice toward the bottom there are some concerns about using this function in the presence of fork. That might not affect us if we wanted to try to use the same `getaddrinfo_a` call.

I did some research on how to do this in Java and pretty much every post just says to use a thread. Obviously that is not ideal.

**Comment by headius** [comment]
D'oh I just read the comment from @mame again:

> Unfortunately, getaddrinfo_a is turned out not to work well with fork. getaddrinfo_a creates some pthread internally, which are not cloned to a child process at fork.

So the Linux way to make getaddrinfo interruptible is exactly the same as all those Java posts: spin up a thread to do the waiting for you. 😂

**Comment by mohamedhafez** [comment]
Ha yeah I'm pretty surprised that there isn't a more standard way of doing this for either C or Java! Anyway, it doesn't seem like timing out DNS lookups is behavior that either JRuby or MRI users can currently expect since `Timeout.timeout` doesn't interrupt the `getaddrinfo` C call in MRI anyway, so my two cents would be that we just get rid of the `Timeout.timeout` anyway since exactly the same current behavior is reproducible with socket timeout options.

The net-http gem says it is compatible back to Ruby 2.6 on rubygems.org, but yeah I get the worry about compatibility, and plus if its fixed in the stdlib for 9.3 then all JRuby users will get it, assuming diverging from MRI on this point isn't a concern. This simple patch I proposed to the net-http gem is passing all tests, I could easily send in a similar pull request to JRuby: https://github.com/ruby/net-http/compare/master...mohamedhafez:patch-1

To exactly mimic current behavior though, we'd have to track the time used up in the DNS lookup, and subtract that from the timeout we give to open the socket. I'm actually doing that in production right now, with a patch that's basically this: https://github.com/jruby/jruby/compare/master...mohamedhafez:patch-1

**Comment by headius** [comment]
If it is possible to get a release of net-http out with this change, compatible with Ruby 2.6 (and with our JRuby changes intact), then I would be in favor of making it a default gem in JRuby 9.3. Otherwise yes, we will want a PR for JRuby.

I will look into the JRuby changes now.

**Comment by mohamedhafez** [comment]
Awesome! For JRuby 9.2 at least, aside from some comments, its just one line removed from net/http.rb I believe. Perhaps the gem would let us put a `unless defined?(JRUBY_VERSION)` in there and make this a whole lot simpler

**Comment by headius** [comment]
Ok one issue with your change is the move to using `Socket.tcp` instead of `TCPSocket.new`. In JRuby, `Socket` is a problematic class because the JDK does not provide any low-level socket support. We are forced to choose from TCP or UDP sockets up front, which makes it difficult to emulate the "build it from scratch" approach of Ruby's `Socket` class.

Socket.tcp could possibly be modified in JRuby to actually use TCPSocket, but that is obviously a larger change.

**Comment by mohamedhafez** [comment]
hmm well `TCPSocket.new` doesn't provide any timeout options unfortunately, and I just pieced this together... I don't actually really know what I'm doing with regards to sockets 😅 

The second patch that uses `Socket.new(:INET, :STREAM)` has been working for me for years, but if there's some issue with it I'm not aware of it's totally understandable if this isn't a good fix. That's actually a lot of why I was bringing this up, to see if what I've been doing is actually acceptable

**Comment by headius** [comment]
@mohamedhafez I am glad to hear it works ok! We basically implement Socket by trying to be lazy in creating the actual JDK objects, so this use case seems to be functional enough.

I understand you want to be able to consider the DNS lookup time as part of the timeout, and you can only do that by doing the lookup separately, and you can only do that by using Socket... so I don't have a better recommendation right now.

**Comment by mohamedhafez** [comment]
Personally I don't really mind about considering the DNS lookup time for my use case, I was just trying to make it mimic exactly the current behavior in hopes of getting a patch accepted. Also just fyi, I hacked net/smtp.rb to use `Socket.tcp` a couple days ago and have been running that in production and haven't seen any issues with that either, although its only been a couple days. It doesn't look to me like there is any way of doing this with `TCPSocket.new`/`TCPSocket.open` without using `Timeout.timeout`, at least with the documented functionality

**Comment by mohamedhafez** [comment]
I don't mean to be on a `Timeout.timeout` crusade over here, if there's problems with the other approaches and in this particular use case of opening a socket all the `handle_interrupt` stuff has been done in a way that it's not dangerous, feel free to close the issue if there's more pressing concerns:)

**Comment by headius** [comment]
@mohamedhafez No no, agreed on all points. I would have liked to see the specific socket classes like TCPSocket get timeout flags, but what we have is what we have. If the existing Socket functionality is working well with your patch we can look at moving forward with it. After all, we do still intend to have 100% support for the Socket class (one way or another) and as you point out there's currently no other way to do it.

**Comment by mohamedhafez** [comment]
Awesome:) so yeah i run a service that for 15,000 users hits an external API once per minute per user, so not huge or anything but the fact that that's been working just fine for years is a pretty good indication that this is feasible. Although maybe taking a look at the source and thinking it through to see if there's anything that could go wrong wouldn't hurt. 

The only weird behavior I can think of is that recently, just starting in the last couple months or so, every now and then when I restart the background process that makes those 15k API requests/minute, about 10 minutes or so after restarting, *sometimes* I get a brief rash of OpenSSL "Socket closed" errors with stack traces back to [this line](https://github.com/mohamedhafez/jruby/blob/5cd7b94fbe890bbe2eef3bdc936ed0baa13d253b/lib/ruby/stdlib/net/http.rb#L1035) (unfortunately for some reason that's where the stack trace ends). After that brief rash everything is fine for days on end. This could be due to any number of things - the external API getting overloaded and misbehaving, recent upgrades to Java 11, maybe the latest version of jruby-openssl, or, just to be extra cautious... perhaps this patch? Like maybe a look at the source of the `Socket` methods I'm calling in the patch to be sure that there's no circumstances where my patch wouldn't actually complete the opening of the socket might be worth it

**Comment by mohamedhafez** [comment]
Actually ignore that last comment - whatever I'm experiencing has nothing to do with this patch, I just restarted the process a couple times *without* the patch, and got the same behavior.

To stay on topic then, I think the simpler patch that uses `Socket.tcp` might be the better one, especially since we're going to need to patch net/smtp.rb, net/pop.rb, net/ftp.rb, net/imap.rb, as well. 

If we want to exactly mimic the current behavior of taking into account DNS lookup time in the timeout, then that fix should be in exactly one place, like `Socket.tcp`, although personally I don't think it's a huge deal. Also in Ruby 2.7 `Socket.tcp` gets a separate `:resolv_timeout` option, so I think it's fine leaving it as is, but your call obviously.

**Comment by headius** [comment]
@mohamedhafez Ideally this should go into the net-http gem and then we can depend on that. I am happy to work with you on getting that patch in and verified on JRuby and released.

If we really can't get this into the net-http gem before 9.3 is released, I see two options:

* Patch it ourselves and not use the gem until 9.4.
* Start using the current net-http gem with the assumption that this will get fixed soon.

**Comment by mohamedhafez** [comment]
sounds like a plan, I sent in https://github.com/ruby/net-http/pull/10 to net-http, hopefully they'll respond

**Comment by mohamedhafez** [comment]
@headius https://github.com/ruby/net-http/pull/10 got merged in!

**Comment by mohamedhafez** [comment]
I've just sent in analogous pull requests to https://github.com/ruby/net-smtp/pull/21, https://github.com/ruby/net-ftp/pull/5, and https://github.com/ruby/net-pop/pull/3, I'm assuming they'll all be accepted since its exactly the same thing I did in https://github.com/ruby/net-http/pull/10. 

If we switch to using those gems for JRuby 9.3.0.0 we'd have eliminated the use of `Timeout.timeout` from the standard libs (with the exception of when we are using SOCKS in net-ftp, since there isn't a `connect_timeout` option available in that case).

**Comment by headius** [comment]
@mohamedhafez I will at least make sure we are running those libraries as default gems, so they can be upgraded independently of JRuby. Ideally we can ship the updated versions, if they are released before 9.3.

Getting rid of all Timeout use in stdlib will be fantastic!
